Improve clarity of try-return-finally-return

toonarmycaptain opened the pull request:
Clarify execution of try-return-finally-return case.
Per discussion https://discuss.python.org/t/pep-601-forbid-return-break-continue-breaking-out-of-finally
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

toonarmycaptain committed:
Improve clarity of try-return-finally-return

Clarify execution in try-return-finally-return case.

toonarmycaptain committed:
Update ACKS

Add my name to ACKS

toonarmycaptain committed:
Update errors.rst

Remove trailing whitespace.

epicfaace commented on Doc/tutorial/errors.rst:
```diff
@@ -393,10 +393,12 @@ statement, whether an exception has occurred or not. When an exception has
 occurred in the :keyword:`!try` clause and has not been handled by an
 :keyword:`except` clause (or it has occurred in an :keyword:`!except` or
 :keyword:`!else` clause), it is re-raised after the :keyword:`finally` clause has
```
```suggestion
:keyword:`!else` clause), it is re-raised after the :keyword:`!finally` clause has
```

epicfaace commented on Doc/tutorial/errors.rst:
```diff
@@ -393,10 +393,12 @@ statement, whether an exception has occurred or not. When an exception has
-when any other clause of the :keyword:`!try` statement is left via a
-:keyword:`break`, :keyword:`continue` or :keyword:`return` statement.  A more
-complicated example::
+been executed.  The :keyword:`!finally` clause is also executed penultimately
```
Reading this for the first time still didn't make this clearer for me. Maybe it would be better to include an example such as https://discuss.python.org/t/pep-601-forbid-return-break-continue-breaking-out-of-finally/2239/3?

toonarmycaptain committed:
Update Doc/tutorial/errors.rst

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

toonarmycaptain committed:
Include example of return in finally returning before return in try

@epicfaace requested including example referred to [on discuss](https://discuss.python.org/t/pep-601-forbid-return-break-continue-breaking-out-of-finally/2239/3?u=toonarmycaptain).

willingc commented on Doc/tutorial/errors.rst:
```diff
@@ -392,11 +392,24 @@ A *finally clause* is always executed before leaving the :keyword:`try`
-when any other clause of the :keyword:`!try` statement is left via a
-:keyword:`break`, :keyword:`continue` or :keyword:`return` statement.  A more
-complicated example::
+:keyword:`!else` clause), it is re-raised after the :keyword:`!finally` clause has
```
Since this is a tutorial, I would recommend being explicit with a direct tone and minimal jargon. Some suggestions for change of this PR:

If a :keyword:`finally` clause is present, the :keyword:`finally` clause will execute as the last task before the :keyword:`try` statement completes. The :keyword:`finally` clause runs whether or not the :keyword:`try` statement produces an exception. The following points discuss more complex cases when an exception occurs:

* If an exception occurs during execution of the :keyword:`!try` clause, the exception may be handled by an :keyword:`except` clause. In all cases, the exception is re-raised after the :keyword:`:`!finally` clause has been executed.

* An exception could occur during execution of an :keyword:`!except` or :keyword:`!else` clause. Again, the exception is re-raised after the :keyword:`:`!finally` clause has been executed. 

* If the :keyword:`!try` statement reaches a :keyword:`break`, :keyword:`continue` or :keyword:`return` statement, the :keyword:`finally` clause will execute just prior to the :keyword:`break`, :keyword:`continue` or :keyword:`return` statement's execution.

* If a :keyword:`finally` clause includes a :keyword:`return` statement, the :keyword:`finally` clause's  :keyword:`return` statement will execute before, and instead of,  the :keyword:`return` statement in a :keyword:`try` clause.

For example:

willingc reviewed (commented):
Hi @toonarmycaptain, thanks for the PR. I have left some suggested wording to help be more explicit and user friendly in tone for this tutorial.

willingc reviewed (changes requested):
Thanks and let me know if you need any clarification on the suggested changes.

toonarmycaptain committed:
Clarify language errors.rst

Clarify language with regard to tone/jargon and formatting [per @willingc](https://github.com/python/cpython/pull/15677/files#r323079139).

toonarmycaptain committed:
Remove errant whitespace.

toonarmycaptain committed:
Add missing backticks around keyword try.

willingc committed:
Fix typo

willingc reviewed (approved):
Thanks @toonarmycaptain. Just waiting for CI to pass.
